File: solo-leveling-notification-service/src/main/java/com/sleepkqq/sololeveling/notification/routing/PriorityBasedRoutingStrategy.java
package com.sleepkqq.sololeveling.notification.routing;

import static com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics.TG_NOTIFICATION_TOPIC;
import static com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics.UI_NOTIFICATION_TOPIC;

import com.sleepkqq.sololeveling.avro.notification.NotificationPriority;
import java.util.Set;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class PriorityBasedRoutingStrategy implements NotificationRoutingStrategy {

  @Override
  public Set<String> getTopics(NotificationPriority priority) {
    return switch (priority) {
      case LOW -> Set.of(UI_NOTIFICATION_TOPIC);
      case MEDIUM, HIGH -> Set.of(UI_NOTIFICATION_TOPIC, TG_NOTIFICATION_TOPIC);
    };
  }
}

File: solo-leveling-notification-service/src/main/kotlin/com/sleepkqq/sololeveling/notification/routing/PriorityBasedRoutingStrategy.kt
package com.sleepkqq.sololeveling.notification.routing

import com.sleepkqq.sololeveling.avro.constants.KafkaTaskTopics
import com.sleepkqq.sololeveling.avro.notification.NotificationPriority
import org.springframework.stereotype.Component

@Suppress("unused")
@Component
class PriorityBasedRoutingStrategy : NotificationRoutingStrategy {

	override fun getTopics(priority: NotificationPriority): Set<String> =
		when (priority) {
			NotificationPriority.LOW -> setOf(KafkaTaskTopics.UI_NOTIFICATION_TOPIC)
			NotificationPriority.MEDIUM, NotificationPriority.HIGH -> setOf(
				KafkaTaskTopics.UI_NOTIFICATION_TOPIC,
				KafkaTaskTopics.TG_NOTIFICATION_TOPIC
			)
		}
}
